### PR TITLE
agent publish refactor

### PR DESCRIPTION
**Pull request by nagworld9**
## Description

Added RSM update verification in agent publish test.

PR includes new pipeline parameter for passing test arguments. Passing all the way from pipeline to each test.
The expected format is comma-separated list of key=value pair like `publihsedVersion=2.10.0.8,hostName=Test`
This is needed especially for agent publish test to know the published version to validate. This new parameter will be useful for other tests when you want to test something dynamically without needing to change tests locally. 

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 1]
```diff
@@ -842,6 +846,18 @@ def _create_test_context(self,) -> AgentTestContext:
                 username=self._user,
                 identity_file=self._identity_file)
 
+    def _get_test_args(self) -> Dict[str, str]:
```
Comment (nagworld9) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:849]: converting list of key=value to dictionary

File: tests_e2e/tests/lib/agent_test.py [hunk 2]
```diff
@@ -49,8 +49,9 @@ class AgentTest(ABC):
     """
     Abstract base class for Agent tests
     """
-    def __init__(self, context: AgentTestContext):
+    def __init__(self, context: AgentTestContext, test_args: Dict[str, str]):
         self._context: AgentTestContext = context
+        self._test_args: Dict[str, str] = test_args
```
Comment (nagworld9) [inline tests_e2e/tests/lib/agent_test.py:54]: These args exposed to all tests

File: tests_e2e/tests/agent_publish/agent_publish.py [hunk 3]
```diff
@@ -86,20 +129,6 @@ def _check_cse(self) -> None:
         )
         custom_script_2_1.assert_instance_view(expected_version="2.1", expected_message=message)
 
-    def get_ignore_error_rules(self) -> List[Dict[str, Any]]:
```
Comment (nagworld9) [inline tests_e2e/tests/agent_publish/agent_publish.py:89]: not needed in this test

File: tests_e2e/tests/agent_update/rsm_update.py [hunk 4]
```diff
@@ -146,64 +140,6 @@ def _prepare_agent(self) -> None:
                               "update-waagent-conf AutoUpdate.UpdateToLatestVersion=y Debug.EnableGAVersioning=y AutoUpdate.GAFamily=Test", use_sudo=True)
         log.info('Successfully updated agent update config \n %s', output)
 
-    @staticmethod
```
Comment (nagworld9) [inline tests_e2e/tests/agent_update/rsm_update.py:149]: moved to common helper

File: tests_e2e/tests/lib/agent_update_helpers.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+def request_rsm_update(requested_version: str, vm: VirtualMachineClient) -> None:
+    """
+    This method is to simulate the rsm request.
+    First we ensure the PlatformUpdates enabled in the vm and then make a request using rest api
+    """
+    if not verify_agent_update_flag_enabled(vm):
+        # enable the flag
+        log.info("Attempting vm update to set the enableVMAgentPlatformUpdates flag")
+        enable_agent_update_flag(vm)
```
Comment (maddieford) [inline tests_e2e/tests/lib/agent_update_helpers.py:61]: Why is this necessary? How is it used vs. our waagent.conf flag

Comment (nagworld9) [reply tests_e2e/tests/lib/agent_update_helpers.py:61]: This different flag need to set in the vm model for CRP/RSM driving RSM updates

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 6]
```diff
@@ -209,6 +211,7 @@ def _initialize(self, environment: Environment, variables: Dict[str, Any], lisa_
         self._environment_name = variables["c_env_name"]
 
         self._test_suites = variables["c_test_suites"]
+        self._test_args = variables["test_args"]
```
Comment (maddieford) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:214]: looks like there may be inconsistency in type here. variables["test_argd"] is str while self._test_args is dict

Comment (nagworld9) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:214]: It's still a string, later we convert to dict. The variables exposed to user in pipeline, runbook or command line is in string in the form of list of comma-separated key=value pair
Example
`publihsedVersion=2.10.0.8,hostName=Test`

Comment (nagworld9) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:214]: oh, you are right. It should be string. I just noticed

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 7]
```diff
@@ -565,6 +568,7 @@ def _execute(self) -> None:
 
                     try:
                         test_context = self._create_test_context()
+                        test_args = self._get_test_args()
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:571]: we should do the conversion in _initialize() and keep the result in self._test_args, to avoid adding more parameters to the execute method

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 8]
```diff
@@ -645,7 +649,7 @@ def _execute_test_suite(self, suite: TestSuiteInfo, test_context: AgentTestConte
 
                         test_success: bool = True
 
-                        test_instance = test.test_class(test_context)
+                        test_instance = test.test_class(test_context, test_args)
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:652]: Rather than passing test_args as a new argument to the test, each test_arg should be added to the text_context as a new property (or if the property already exists in the context, the just set its value).

Then AgentPublishTest.run_from_command_line should read the new parameters from the command line and add them to the test context as well. That is, it would parse the publishedVersion argument, instead of the test_args argument

**Review by narrieta — COMMENTED** [review]
Added some comments, could you also describe how you are planning to planning to pass the arguments to the publish test from the pipeline? If the new argument is required, we should probably create a new pipeline.yml for the publish pipeline.

**Comment by nagworld9** [comment]
> Added some comments, could you also describe how you are planning to planning to pass the arguments to the publish test from the pipeline? If the new argument is required, we should probably create a new pipeline.yml for the publish pipeline.

My idea is to it would benefit for other test_suites(for dev runs etc) if we added to main pipeline because the internal implementation is going to be same regardless. 

When we run the pipeline, one need to pass like list of key=value pair in the pipeline arg. This converted to dict and added to agenttestcontext, so that all the tests access to dict. Whichever test needed, they read the keys to get value.

For publish test when I run the pipeline with agent_publish as test_suite, I pass publishedVersion={version}, the publish test reads this key publihsedVersion

**Comment by narrieta** [comment]
> My idea is to it would benefit for other test_suites(for dev runs etc) if we added to main pipeline because the internal implementation is going to be same regardless.

Yes, I agree that it is useful and can be added as an optional parameter in the main pipeline

> For publish test when I run the pipeline with agent_publish as test_suite, I pass publishedVersion={version}, the publish test reads this key publihsedVersion

If this parameter is required by the publish test, then maybe should have a pipeline just for publish,

**Comment by nagworld9** [comment]
> > My idea is to it would benefit for other test_suites(for dev runs etc) if we added to main pipeline because the internal implementation is going to be same regardless.
> 
> Yes, I agree that it is useful and can be added as an optional parameter in the main pipeline
> 
> > For publish test when I run the pipeline with agent_publish as test_suite, I pass publishedVersion={version}, the publish test reads this key publihsedVersion
> 
> If this parameter is required by the publish test, then maybe should have a pipeline just for publish,

Not sure I followed here. I wouldn't send this as parameter, rather I send this parameter as value in the new the test_arg parameter. That way I don't have deal with separate code path just for agent publish

**Comment by nagworld9** [comment]
@narrieta Updated my pr. I added todo for reading it from cmd line. I couldn't get the flow right, maybe I'm missing something and also, I have different thought like we can directly use argparser for reading test specific arguments in the test itself.

File: tests_e2e/tests/lib/agent_test.py [hunk 9]
```diff
@@ -73,6 +73,8 @@ def run_from_command_line(cls):
         """
         Convenience method to execute the test when it is being invoked directly from the command line (as opposed as
         being invoked from a test framework or library.)
+
+        Todo: Need to implement for reading test specific arguments from command line
```
Comment (narrieta) [inline tests_e2e/tests/lib/agent_test.py:77]: Let's use capitals: TODO. Then some tools highlight that comment

**Review by narrieta — APPROVED** [review]

**Review by maddieford — APPROVED** [review]
